packages/snaps-execution-environments/src/webview/ProxyMessageStream.ts

```diff
@@ -0,0 +1,95 @@
+import type { PostMessageEvent } from '@metamask/post-message-stream';
+import { BasePostMessageStream } from '@metamask/post-message-stream';
+import { isValidStreamMessage } from '@metamask/post-message-stream/dist/utils';
+import { base64ToBytes, bytesToString } from '@metamask/utils';
+
+type ProxyMessageStreamArgs = {
+  name: string;
+  target: string;
+  targetOrigin?: string;
+  targetWindow?: Window;
+};
+
+export class ProxyMessageStream extends BasePostMessageStream {
+  #name;
+
+  #target;
+
+  #targetOrigin;
+
+  #targetWindow;
+
+  /**
+   * Creates a stream for communicating with other streams across the same or
+   * different `window` objects.
+   *
+   * @param args - Options bag.
+   * @param args.name - The name of the stream. Used to differentiate between
+   * multiple streams sharing the same window object. child:WebView
+   * @param args.target - The name of the stream to exchange messages with. parent:rnside
+   * @param args.targetOrigin - The origin of the target. Defaults to
+   * `location.origin`, '*' is permitted. WebView's main Frame location
+   * @param args.targetWindow - The window object of the target stream. Defaults
+   * to `window`. WebView's main Frame Window
+   */
+
+  constructor({
+    name,
+    target,
+    targetOrigin = window.location.origin,
+    targetWindow = window,
+  }: ProxyMessageStreamArgs) {
+    super();
+
+    this.#name = name;
+    this.#target = target;
+    this.#targetOrigin = targetOrigin;
+    this.#targetWindow = targetWindow;
+
+    this._onMessage = this._onMessage.bind(this);
+
+    window.addEventListener(
+      'message',
+      (event: MessageEvent<PostMessageEvent>) => this._onMessage(event as any),
+      false,
+    );
+
+    this._handshake();
+  }
+
+  /**
+   * Webview needs to receive strings only on postMessage. That's the main difference between this and the original window post message stream.
+   * Reference: https://github.com/react-native-webview/react-native-webview/blob/master/docs/Guide.md?plain=1#L471
+   */
+
+  protected _postMessage(data: unknown): void {
+    this.#targetWindow.ReactNativeWebView.postMessage(
```

Comment:
```suggestion
    this.#targetWindow.postMessage(
```
Can we pass in the `ReactNativeWebView` instead of the global window?